platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+up to match the new version of the form.
+
+A `dataMigrations` prop may be passed to the `SaveInProgress` component.
+`dataMigrations` must be an array of functions. Each function will be called
+with the form data and metadata, either from the API request or a previous
+migration function. Metadata will include:
+- Form version when the data was saved
+- The page URL when the form was saved
+
+Each function must return the new form data and metadata. If multiple migration
+functions must be run, the result of an earlier migration will be passed to the
+next migration in the array.
+
+Migrations are run during the `loadForm` function. It will check the form
+version in the metadata against the length of the `dataMigrations` array. If the
+latter is greater, all migrations after the saved form version will be called in
+order.
+    
+#### JSON schema validation
+To ensure the data sent from the front end matches data expected on the back
+end, the client and API must share a common schema. Forms on VA.gov have used
+the [JSON schema](https://json-schema.org) specification for this.
+
+While Formik has [built-in support for
+`Yup`](https://formik.org/docs/guides/validation#validationschema) schemas,
+these are JavaScript-only and not easily shareable with the API. As such, we
+will continue to support JSON schema by writing a function which can be used
+while creating the [validate
+function](https://formik.org/docs/guides/validation#validate) for each page or
+field.
+
+The JSON schema validation function will accept a JSON schema and return a
+closure which can be used to validate individual fields against the schema. This
+closure will also have chainable functions which may be used to override the default
+error messages.
+
+Formulate will use the [`jsonschema`](https://github.com/tdegrunt/jsonschema)
+package to perform the JSON schema validation due to its
+[popularity](https://www.npmjs.com/package/jsonschema), [small
+size](https://bundlephobia.com/result?p=jsonschema@1.4.0), and [parseable
+result](https://github.com/tdegrunt/jsonschema#results).
+
+Each [JSON schema validation
+keyword](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6)
+will have a default error message and a corresponding function to override that
+default. This function will accept either a string or a function that's passed
+the value of the field as a parameter. (See below for an example.)
+
+##### Usage
+**Setup:**
+```js
+// schema-validation.js
+import loadSchema from 'formulate/schema-validation';
+import jsonSchema from '@department-of-veterans-affairs/vets-json-schema/my-form-schema.json';
+
+const fromSchema = loadSchema(jsonSchema);
+
+export fromSchema;
+```
+
+**Single validate function for the page:**
+```js
+import fromSchema from '../schema-validation';
+
+const validate = values => {
+  const errors = {};
+  
+  const passwordErrorMessage = fromSchema('password'), // Path to schema
```

Comment:
Is that trailing comma necessary, or am I misunderstanding what this syntax is doing?

Reply:
Oh, nope. That's a typo. Good catch. :male_detective: